### PR TITLE
Fix compact

### PR DESCRIPTION
**Pull request by small-turtle-1**
### What problem does this PR solve?

1. Segment now does not save `compacting_ts`/`no_delete_ts`/`deprecate_ts` because multi version is not used when compaction. Segment now has status instead.
2. Fix bug: compact thread-safe. 
    1. Whether delete txn on compacting segment will commit is decided in `CheckDeleteConflict` called by worker thread. It throws a recoverable exception if this segment is set "kNoDelete" or "kDeprecate". Or it will add reference count on segment for each txn. And when delete txn is commit or rollbacked, the ref cnt is minused. When commit, the delete record will be append to compacting task if exists.
    2. When compaction task is ready to commit. It will excute the delete record during compaction. Which set "NoDelete" to prehibit new delete txn and then wait for txn ref cnt to zero which means no delete record will be appended. 
3. Fix warning in issue.
4. Fix bug: catalog flush bg task, which is not thread safe originally. Now, flush bg task add lock on "delta catalog entry" and pick the flushable entry. And Txn local "delta catalog entries" should add lock too, because it may accessed by multiple worker threads.
5. Fix bug: Manual compaction and auto compaction will not interfere each other now. Manual compaction set `TableEntry::compaction_alg_` to disable(may wait) and set enable after commit or rollback. Auto compaction will not trigger when `TableEntry::compaction_alg_` is disable.

Issue link: #572
#562 

### Type of change

- [x] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [x] Test cases
- [ ] Other (please describe):

### CODE REVIEW

File: src/storage/bg_task/compact_segments_task.cpp [hunk 1]
```diff
@@ -117,45 +120,72 @@ void CompactSegmentsTask::Execute() {
...
+        if (to_compact_segments.empty()) {
+            return;
+        }
+        {
+            String ss;
+            ss += "Compacting segments: ";
+            for (auto *segment : to_compact_segments) {
+                ss += std::to_string(segment->segment_id()) + " ";
+            }
+            LOG_INFO(fmt::format("Compacting segments: {}", ss));
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cpp:141]: LOG_TRACE

File: src/storage/bg_task/compact_segments_task.cpp [hunk 2]
```diff
@@ -117,45 +120,72 @@ void CompactSegmentsTask::Execute() {
...
             }
             break;
         }
         case CompactSegmentsTaskType::kCompactPickedSegments: {
             // the picked segments are already in table ref
-            Vector<SegmentEntry *> to_compact_segments = table_ref_->block_index_->segments_; // copy the vector
+            Vector<SegmentEntry *> to_compact_segments;
+            for (auto *segment : table_ref_->block_index_->segments_) {
+                if (!segment->TrySetCompacting(this)) {
+                    UnrecoverableError("Assert");
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cpp:174]: Give a meaningful error message.

File: src/storage/meta/entry/segment_entry.cpp [hunk 3]
```diff
@@ -14,6 +14,8 @@
 
 module;
 
+#include <bits/std_thread.h>
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cpp:17]: #include <thread>?

File: src/storage/meta/entry/segment_entry.cpp [hunk 4]
```diff
@@ -99,73 +102,83 @@ SharedPtr<SegmentEntry> SegmentEntry::NewReplayCatalogSegmentEntry(const TableEn
...
     return true;
 }
 
-void SegmentEntry::SetNoDelete(TxnTimeStamp no_delete_ts) {
+void SegmentEntry::SetNoDelete() {
+
     std::unique_lock lock(rw_locker_);
-    no_delete_ts_ = no_delete_ts;
+    if (status_ != SegmentStatus::kCompacting) {
+        UnrecoverableError("Assert.");
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cpp:130]: Error message.

File: src/storage/meta/entry/segment_entry.cpp [hunk 5]
```diff
@@ -99,73 +102,83 @@ SharedPtr<SegmentEntry> SegmentEntry::NewReplayCatalogSegmentEntry(const TableEn
...
 
-void SegmentEntry::SetDeprecated(TxnTimeStamp deprecate_ts) {
+void SegmentEntry::SetDeprecated() {
     std::unique_lock lock(rw_locker_);
-    deprecate_ts_ = deprecate_ts;
-
-    // this is not necessary, but for safety
-    compact_task_ = nullptr;
+    if (status_ != SegmentStatus::kNoDelete) {
+        UnrecoverableError("Assert.");
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cpp:140]: Error message

File: src/storage/meta/entry/segment_entry.cpp [hunk 6]
```diff
@@ -99,73 +102,83 @@ SharedPtr<SegmentEntry> SegmentEntry::NewReplayCatalogSegmentEntry(const TableEn
...
 void SegmentEntry::RollbackCompact() {
     std::unique_lock lock(rw_locker_);
-    compacting_ts_ = UNCOMMIT_TS;
-    no_delete_ts_ = UNCOMMIT_TS;
-    deprecate_ts_ = UNCOMMIT_TS;
-
-    // this is not necessary, but for safety
-    compact_task_ = nullptr;
+    if (status_ != SegmentStatus::kNoDelete) {
+        UnrecoverableError("Assert.");
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cpp:148]: Error message

File: src/storage/meta/entry/segment_entry.cppm [hunk 7]
```diff
@@ -175,22 +179,23 @@ private:
...
-    TxnTimeStamp compacting_ts_{UNCOMMIT_TS};   // Indicate the commit_ts which start compacting this SegmentEntry
-    TxnTimeStamp no_delete_ts_{UNCOMMIT_TS};    // Indicate the commit_ts which prehibit delete in this SegmentEntry
-    TxnTimeStamp deprecate_ts_{UNCOMMIT_TS};    // Indicate the commit_ts which deprecate this SegmentEntry
 
     Vector<SharedPtr<BlockEntry>> block_entries_{};
 
     CompactSegmentsTask *compact_task_{};
+    SegmentStatus status_;
+
+    std::condition_variable_any txn_num_cv_{};
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cppm:197]: Why not use std::condition_variable?

File: src/storage/meta/entry/table_entry.cpp [hunk 8]
```diff
@@ -349,6 +352,7 @@ Status TableEntry::CommitCompact(TransactionID txn_id, TxnTimeStamp commit_ts, c
             break;
         }
         default:
+            break;
```
Comment (JinHai-CN) [inline src/storage/meta/entry/table_entry.cpp:355]: Error message? Unrecoverable Error?

File: src/storage/meta/entry/table_entry.cpp [hunk 9]
```diff
@@ -366,6 +370,7 @@ Status TableEntry::RollbackCompact(TransactionID txn_id, TxnTimeStamp commit_ts,
...
```
Comment (JinHai-CN) [inline src/storage/meta/entry/table_entry.cpp:373]: Error message? Unrecoverable Error?

File: src/storage/wal/wal_manager.cpp [hunk 10]
```diff
@@ -728,7 +728,11 @@ void WalManager::WalCmdCompactReplay(const WalCmdCompact &cmd, TransactionID txn
 
     for (const SegmentID segment_id : cmd.deprecated_segment_ids_) {
         auto *segment_entry = table_entry->GetSegmentByID(segment_id, commit_ts);
-        segment_entry->SetDeprecated(commit_ts);
+        if (!segment_entry->TrySetCompacting(nullptr)) { // fake set because check
+            UnrecoverableError("Assert.");
```
Comment (JinHai-CN) [inline src/storage/wal/wal_manager.cpp:732]: Error message

File: src/storage/meta/entry/segment_entry.cppm (hunk 7 above)
Comment (small-turtle-1) [reply src/storage/meta/entry/segment_entry.cppm:197]: condition_variable cannot wait on std::shared_mutex (only work on std::mutex)

File: src/storage/meta/entry/segment_entry.cpp [hunk 11]
```diff
@@ -235,59 +246,87 @@ u64 SegmentEntry::AppendData(TransactionID txn_id, AppendState *append_state_ptr
...
+                }
+            }
+            compact_task_->AddToDelete(segment_id_, std::move(segment_offsets));
+        }
         {
-            std::unique_lock<std::shared_mutex> lck(this->rw_locker_);
-            block_entry = block_entries_.at(block_id).get();
+            delete_txns_.erase(txn_id);
+            if (delete_txns_.empty()) { // == 0 when replay
+                txn_num_cv_.notify_one();
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cpp:307]: no_delete_complete_cv_;

File: src/storage/meta/entry/segment_entry.cpp [hunk 12]
```diff
@@ -235,59 +246,87 @@ u64 SegmentEntry::AppendData(TransactionID txn_id, AppendState *append_state_ptr
...
+    {
+        std::unique_lock lck(this->rw_locker_);
+        this->max_row_ts_ = std::max(this->max_row_ts_, commit_ts);
+        if (this->first_delete_ts_ == UNCOMMIT_TS) {
+            this->first_delete_ts_ = commit_ts;
+        }
+        if (status_ == SegmentStatus::kDeprecated) {
+            UnrecoverableError("Assert: Should not commit delete to deprecated segment.");
+        }
+        if (compact_task_ != nullptr) {
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cpp:291]: compact_task_ should be in table_entry.

File: src/storage/meta/entry/segment_entry.cpp [hunk 13]
```diff
@@ -99,73 +100,81 @@ SharedPtr<SegmentEntry> SegmentEntry::NewReplayCatalogSegmentEntry(const TableEn
...
 void SegmentEntry::RollbackCompact() {
     std::unique_lock lock(rw_locker_);
-    compacting_ts_ = UNCOMMIT_TS;
-    no_delete_ts_ = UNCOMMIT_TS;
-    deprecate_ts_ = UNCOMMIT_TS;
-
-    // this is not necessary, but for safety
-    compact_task_ = nullptr;
+    if (status_ != SegmentStatus::kNoDelete) {
+        UnrecoverableError("Assert: Rollbacked segment should be kNoDelete.");
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cpp:145]: kNoDelete->NoDelete

Comment (JinHai-CN) [reply src/storage/meta/entry/segment_entry.cpp:145]: No Delete state.

File: src/storage/meta/entry/segment_entry.cpp [hunk 14]
```diff
@@ -185,13 +194,15 @@ void SegmentEntry::AppendBlockEntry(UniquePtr<BlockEntry> block_entry) {
     block_entries_.emplace_back(std::move(block_entry));
 }
 
+// One writer
 u64 SegmentEntry::AppendData(TransactionID txn_id, AppendState *append_state_ptr, BufferManager *buffer_mgr, Txn *txn) {
-    std::unique_lock<std::shared_mutex> lck(this->rw_locker_); // FIXME: lock scope to large
-    if (this->sealed()) {
+    std::unique_lock lck(this->rw_locker_); // FIXME: lock scope too large
+    if (this->status_ != SegmentStatus::kUnsealed) {
         UnrecoverableError("AppendData to sealed segment");
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cpp:201]: Update error message.

File: src/storage/meta/entry/segment_entry.cpp [hunk 15]
```diff
@@ -235,59 +246,87 @@ u64 SegmentEntry::AppendData(TransactionID txn_id, AppendState *append_state_ptr
     return total_copied;
 }
 
+// One writer
 void SegmentEntry::DeleteData(TransactionID txn_id, TxnTimeStamp commit_ts, const HashMap<BlockID, Vector<BlockOffset>> &block_row_hashmap) {
     for (const auto &[block_id, delete_rows] : block_row_hashmap) {
         BlockEntry *block_entry = nullptr;
         {
-            std::unique_lock<std::shared_mutex> lck(this->rw_locker_);
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cpp:242]: Use previous implementation and change locker to shared_lock

File: src/storage/meta/entry/segment_entry.cpp [hunk 16]
```diff
@@ -235,59 +246,87 @@ u64 SegmentEntry::AppendData(TransactionID txn_id, AppendState *append_state_ptr
...
+        std::unique_lock lck(this->rw_locker_);
+        if (this->status_ != SegmentStatus::kUnsealed) {
+            UnrecoverableError("Assert: Should not commit append to sealed segment.");
+        }
         if (this->min_row_ts_ == UNCOMMIT_TS) {
             this->min_row_ts_ = commit_ts;
         }
-        block_entry = this->block_entries_[block_id];
+        this->max_row_ts_ = std::max(this->max_row_ts_, commit_ts);
+        block_entry = this->block_entries_.at(block_id).get();
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cpp:274]: Exception?

File: src/storage/meta/entry/segment_entry.cpp [hunk 17]
```diff
@@ -235,59 +246,87 @@ u64 SegmentEntry::AppendData(TransactionID txn_id, AppendState *append_state_ptr
...
         {
-            std::unique_lock<std::shared_mutex> lck(this->rw_locker_);
-            block_entry = block_entries_.at(block_id).get();
+            delete_txns_.erase(txn_id);
+            if (delete_txns_.empty()) { // == 0 when replay
+                txn_num_cv_.notify_one();
+            }
+        }
+        for (const auto &[block_id, delete_rows] : block_row_hashmap) {
+            BlockEntry *block_entry = block_entries_.at(block_id).get();
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cpp:311]: raise exception

File: src/storage/meta/entry/segment_entry.cpp [hunk 18]
```diff
@@ -235,59 +246,87 @@ u64 SegmentEntry::AppendData(TransactionID txn_id, AppendState *append_state_ptr
...
             if (block_entry == nullptr) {
                 UnrecoverableError(fmt::format("The segment doesn't contain the given block: {}.", block_id));
             }
             if (delete_rows.size() > block_entry->row_capacity()) {
                 UnrecoverableError("Delete rows exceed block capacity");
             }
             DecreaseRemainRow(delete_rows.size());
-            this->first_delete_ts_ = std::min(this->first_delete_ts_, commit_ts);
+            delete_blocks.push_back(block_entry);
```
Comment (JinHai-CN) [inline src/storage/meta/entry/segment_entry.cpp:319]: All vector push_back can be reviewed if vector::reserve can be used.

File: src/storage/meta/entry/segment_entry.cpp (hunk 12 above)
Comment (small-turtle-1) [reply src/storage/meta/entry/segment_entry.cpp:291]: Change much. Next pr.

**Pull request merged**